index.bs

```diff
@@ -612,6 +655,101 @@ A mapping Web application uses the device's orientation to correctly align the m
 * The specification must be agnostic to the underlying sources of orientation and motion data.
 * The specification must use the existing DOM event framework.
 
+Automation {#automation}
+==========
+
+This specification can pose a challenge to test authors, as the events defined here depend on the presence of physical hardware whose readings cannot be easily controlled.
+
+To address this challenge, this document builds upon the [[WEBDRIVER2]] <a>extension commands</a> and infrastructure laid out by [[GENERIC-SENSOR#automation]]. This was chosen over the option of developing completely new and independent infrastructure with separate <a>extension commands</a> because there is significant overlap between the two specifications: not only does testing the [[GENERIC-SENSOR]] specification present similar challenges, but many derived APIs (e.g. [[GYROSCOPE]]) obtain and provide similar information.
+
+This specification only requires implementations to support the [[GENERIC-SENSOR#automation]] section of the [[GENERIC-SENSOR]] specification, not its interfaces and events.
+
+Device Orientation Automation {#device-orientation-automation}
+-----------------------------
+
+Automation support for the <a event><code>deviceorientation</code></a> event is built upon virtual sensors that represent accelerometers, gyroscopes and, optionally, magnetometers.
+
+Orientation data retrieved from the platform by the user agent comes from accelerometers, gyroscopes and, optionally, magnetometers. Contrary to motion data, however, these lower-level readings must be transformed into Euler angles in the formation described in [[#device-orientation-model]]. Furthermore, the platform might provide extra APIs to the user agent that already perform some of those conversions from raw acceleration and rotation data.
+
+Therefore, instead of requiring implementations (and automation users) to provide orientation readings via lower-level virtual sensors which use different units of measurement, this specification defines extra <a>virtual sensor types</a> for relative and orientation data in the format used by this specification.
+
+<h4 id="parse-orientation-data-reading-algorithm">Parse orientation data reading algorithm</h4>
+
+<div algorithm>
+To perform the <dfn>parse orientation data reading</dfn> algorithm, given a JSON {{Object}} <var>parameters</var>:
+
+    1. Let <var>alpha</var> be the result of invoking <a>get a property</a> from <var>parameters</var> with "<code>alpha</code>".
+    1. If <var>alpha</var> is not a {{Number}}, or its value is <strong>NaN</strong>, +∞, or −∞, return <strong>undefined</strong>.
```

Comment:
@reillyeon btw I wonder if we should support null for alpha, beta and/or gamma.

The other sensors do not support null values as the specs themselves do not seem to support them, but since this is a new kind of sensor type that is only used here, it could in theory be done. The other option is to just rely on the "can provide readings" flag being false and have the 3 readings set to null as a side-effect (as we've been doing with DeviceMotionEvent and its web tests).

Reply:
I would rely on "can provide readings".

Reply:
Ack.